### ISSUE TITLE
UnsupportedProtocolError on using Artifactory Docker registry with port

### ISSUE THREAD

**Issue by ps-jay** [renovatebot/renovate#6904]
**What Renovate type are you using?**

Self-hosted Renovate (v22.7.0)

Self-hosted Artifactory

**Describe the bug**

`UnsupportedProtocolError` on trying to access the self-hosted Artifactory Docker registry.

Looks to me like Renovate is confusing the registry domain name with the protocol part of the URL.  Parsing error of the `FROM` line in the Dockerfile maybe?  I think it should have assumed the protocol was HTTPS.

**Relevant debug logs**

```
build	04-Aug-2020 00:27:46	DEBUG: Matched 1 file(s) for manager dockerfile: Dockerfile (repository=dbi/centos7-java-base)
build	04-Aug-2020 00:27:46	DEBUG: Found dockerfile package files (repository=dbi/centos7-java-base)
build	04-Aug-2020 00:27:46	DEBUG: Found 1 package file(s) (repository=dbi/centos7-java-base)
build	04-Aug-2020 00:27:46	 INFO: Dependency extraction complete (repository=dbi/centos7-java-base)
build	04-Aug-2020 00:27:46	       "baseBranch": "master",
build	04-Aug-2020 00:27:46	       "stats": {
build	04-Aug-2020 00:27:46	         "managers": {"dockerfile": {"fileCount": 1, "depCount": 1}},
build	04-Aug-2020 00:27:46	         "total": {"fileCount": 1, "depCount": 1}
build	04-Aug-2020 00:27:46	       }
build	04-Aug-2020 00:27:46	 WARN: Error obtaining docker token (repository=dbi/centos7-java-base)
build	04-Aug-2020 00:27:46	       "registry": "artifactory.example.org:6555/",
build	04-Aug-2020 00:27:46	       "dockerRepository": "base/centos7_example_base",
build	04-Aug-2020 00:27:46	       "err": {
build	04-Aug-2020 00:27:46	         "name": "UnsupportedProtocolError",
build	04-Aug-2020 00:27:46	         "message": "Unsupported protocol \"artifactory.example.org:\"",
build	04-Aug-2020 00:27:46	         "stack": "UnsupportedProtocolError: Unsupported protocol \"artifactory.example.org:\"\n    at normalizeArguments (/usr/src/app/node_modules/got/dist/source/core/index.js:452:23)\n    at normalizeArguments (/usr/src/app/node_modules/got/dist/source/as-promise/core.js:34:31)\n    at Function.got [as default] (/usr/src/app/node_modules/got/dist/source/create.js:104:39)\n    at Http.request (/usr/src/app/node_modules/renovate/lib/util/http/index.ts:128:28)\n    at Http.get (/usr/src/app/node_modules/renovate/lib/util/http/index.ts:144:17)\n    at getAuthHeaders (/usr/src/app/node_modules/renovate/lib/datasource/docker/index.ts:142:41)\n    at getTags (/usr/src/app/node_modules/renovate/lib/datasource/docker/index.ts:393:27)\n    at Object.getReleases (/usr/src/app/node_modules/renovate/lib/datasource/docker/index.ts:589:16)\n    at getRegistryReleases (/usr/src/app/node_modules/renovate/lib/datasource/index.ts:54:15)\n    at fetchReleases (/usr/src/app/node_modules/renovate/lib/datasource/index.ts:186:15)\n    at Object.getPkgReleases (/usr/src/app/node_modules/renovate/lib/datasource/index.ts:248:7)\n    at Object.lookupUpdates (/usr/src/app/node_modules/renovate/lib/workers/repository/process/lookup/index.ts:152:30)\n    at fetchDepUpdates (/usr/src/app/node_modules/renovate/lib/workers/repository/process/fetch.ts:44:26)\n    at /usr/src/app/node_modules/p-map/index.js:57:22"
build	04-Aug-2020 00:27:46	       }
build	04-Aug-2020 00:27:46	DEBUG: Failed to get authHeaders for getTags lookup (repository=dbi/centos7-java-base)
build	04-Aug-2020 00:27:46	DEBUG: Failed to look up dependency artifactory.example.org:6555/base/centos7_example_base (repository=dbi/centos7-java-base, packageFile=Dockerfile, dependency=artifactory.example.org:6555/base/centos7_example_base)
build	04-Aug-2020 00:27:46	DEBUG: Package releases lookups complete (repository=dbi/centos7-java-base)
build	04-Aug-2020 00:27:46	       "baseBranch": "master"
build	04-Aug-2020 00:27:46	DEBUG: packageFiles with updates (repository=dbi/centos7-java-base)
build	04-Aug-2020 00:27:46	       "config": {
build	04-Aug-2020 00:27:46	         "dockerfile": [
build	04-Aug-2020 00:27:46	           {
build	04-Aug-2020 00:27:46	             "packageFile": "Dockerfile",
build	04-Aug-2020 00:27:46	             "manager": "dockerfile",
build	04-Aug-2020 00:27:46	             "deps": [
build	04-Aug-2020 00:27:46	               {
build	04-Aug-2020 00:27:46	                 "depName": "artifactory.example.org:6555/base/centos7_example_base",
build	04-Aug-2020 00:27:46	                 "currentValue": "2020.8.1",
build	04-Aug-2020 00:27:46	                 "replaceString": "artifactory.example.org:6555/base/centos7_example_base:2020.8.1",
build	04-Aug-2020 00:27:46	                 "autoReplaceStringTemplate": "{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
build	04-Aug-2020 00:27:46	                 "datasource": "docker",
build	04-Aug-2020 00:27:46	                 "depType": "final",
build	04-Aug-2020 00:27:46	                 "depIndex": 0,
build	04-Aug-2020 00:27:46	                 "updates": [],
build	04-Aug-2020 00:27:46	                 "warnings": [
build	04-Aug-2020 00:27:46	                   {
build	04-Aug-2020 00:27:46	                     "depName": "artifactory.example.org:6555/base/centos7_example_base",
build	04-Aug-2020 00:27:46	                     "message": "Failed to look up dependency artifactory.example.org:6555/base/centos7_example_base"
build	04-Aug-2020 00:27:46	                   }
build	04-Aug-2020 00:27:46	                 ]
build	04-Aug-2020 00:27:46	               }
build	04-Aug-2020 00:27:46	             ]
build	04-Aug-2020 00:27:46	           }
build	04-Aug-2020 00:27:46	         ]
build	04-Aug-2020 00:27:46	       }
```

(sorry for the timestamps from the Bamboo log)

**To Reproduce**

Dockerfile looks like this:

```
FROM artifactory.example.org:6555/base/centos7_example_base:2020.8.1
RUN ln -sf /usr/share/zoneinfo/Australia/Melbourne /etc/localtime

# Add Example CA
COPY example-*.crt /tmp/
COPY svc_trs* /etc/

COPY log4j2.xml /opt/log4j/
COPY log4j2.xml /opt/log4j/log4j2_json_sentry_file.xml
COPY log4j2.xml /opt/log4j/log4j2_json_sentry.xml
COPY log4j2.xml /opt/log4j/log4j2_json.xml
COPY log4j2.xml /opt/log4j/log4j2_sentry.xml

RUN cd /opt/newrelic; curl -L https://download.newrelic.com/newrelic/java-agent/newrelic-agent/current/newrelic-agent.jar -o newrelic-agent.jar
COPY newrelic-ignore-health-checks.xml /opt/newrelic/extensions/
```

**Additional context**

We upgraded from Renovate v19.239.11 to Renovate v22.7.0, which exposed the fault.

A successful run the previous day looked like this:

```
build	03-Aug-2020 11:41:07	 INFO: Dependency extraction complete (repository=dbi/centos7-java-base)
build	03-Aug-2020 11:41:07	       "baseBranch": "master",
build	03-Aug-2020 11:41:07	       "stats": {
build	03-Aug-2020 11:41:07	         "managers": {"dockerfile": {"fileCount": 1, "depCount": 1}},
build	03-Aug-2020 11:41:07	         "total": {"fileCount": 1, "depCount": 1}
build	03-Aug-2020 11:41:07	       }
build	03-Aug-2020 11:41:07	DEBUG: getLabels(artifactory.example.org:6555/, base/centos7_example_base, latest) (repository=dbi/centos7-java-base)
build	03-Aug-2020 11:41:07	DEBUG: Package releases lookups complete (repository=dbi/centos7-java-base)
build	03-Aug-2020 11:41:07	       "baseBranch": "master"
build	03-Aug-2020 11:41:07	DEBUG: packageFiles with updates (repository=dbi/centos7-java-base)
build	03-Aug-2020 11:41:07	       "config": {
build	03-Aug-2020 11:41:07	         "dockerfile": [
build	03-Aug-2020 11:41:07	           {
build	03-Aug-2020 11:41:07	             "packageFile": "Dockerfile",
build	03-Aug-2020 11:41:07	             "manager": "dockerfile",
build	03-Aug-2020 11:41:07	             "deps": [
build	03-Aug-2020 11:41:07	               {
build	03-Aug-2020 11:41:07	                 "depName": "artifactory.example.org:6555/base/centos7_example_base",
build	03-Aug-2020 11:41:07	                 "currentValue": "2020.5.3",
build	03-Aug-2020 11:41:07	                 "replaceString": "artifactory.example.org:6555/base/centos7_example_base:2020.5.3",
build	03-Aug-2020 11:41:07	                 "autoReplaceStringTemplate": "{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
build	03-Aug-2020 11:41:07	                 "datasource": "docker",
build	03-Aug-2020 11:41:07	                 "depType": "final",
build	03-Aug-2020 11:41:07	                 "depIndex": 0,
build	03-Aug-2020 11:41:07	                 "updates": [
build	03-Aug-2020 11:41:07	                   {
build	03-Aug-2020 11:41:07	                     "fromVersion": "2020.5.3",
build	03-Aug-2020 11:41:07	                     "toVersion": "2020.8.1",
build	03-Aug-2020 11:41:07	                     "newValue": "2020.8.1",
build	03-Aug-2020 11:41:07	                     "newMajor": 2020,
build	03-Aug-2020 11:41:07	                     "newMinor": 8,
build	03-Aug-2020 11:41:07	                     "updateType": "minor",
build	03-Aug-2020 11:41:07	                     "isSingleVersion": true
build	03-Aug-2020 11:41:07	                   }
build	03-Aug-2020 11:41:07	                 ],
build	03-Aug-2020 11:41:07	                 "warnings": [],
build	03-Aug-2020 11:41:07	                 "sourceUrl": null,
build	03-Aug-2020 11:41:07	                 "dockerRegistry": "artifactory.example.org:6555/",
build	03-Aug-2020 11:41:07	                 "dockerRepository": "base/centos7_example_base"
build	03-Aug-2020 11:41:07	               }
build	03-Aug-2020 11:41:07	             ]
build	03-Aug-2020 11:41:07	           }
build	03-Aug-2020 11:41:07	         ]
build	03-Aug-2020 11:41:07	       }
```

**Comment by zharinov** [comment]
@ps-jay I was trying to reproduce with the provided example, that's what I've got:
``` 
WARN: Error obtaining docker token (repository=zharinov/renovate-test-docker)
"registry": "https://artifactory.example.org:6555",
"dockerRepository": "base/centos7_example_base",
"err": {
  "name": "RequestError",
  "code": "ECONNREFUSED",
  ...
}
```
In my case, the protocol seems to be correctly defined, though getting another error (which should not be the case for real registry). Looks like I managed to figure out the underlying problem anyway, but can you please clarify your settings for docker in `renovate.json`?

**Comment by ps-jay** [comment]
Hi,

Thanks for having a look.

example.org is a fake domain sorry. Our Artifactory instance is 'behind the
firewall'.

I'll post our config.js later.


Regards,
  Phil

**Comment by rarkins** [comment]
Please try to avoid posting incomplete or "fake" reproduction reproductions as it drains our time

**Comment by ps-jay** [comment]
I see that.  Please accept my apologies.

I'll reproduce in a way that you will be able to as well, and update this issue.

**Comment by ps-jay** [comment]
I've been trying to find a public, but non-Docker Hub, repo to reproduce this.  I haven't succeeded in that yet.

However, I've been trying to narrow down to where the problem began.
I think it relates to the update of the `got` module.

v21.33.15 works okay.
v22.0.3 exhibits the issue (22.0.3 is the first 22.0.x release on Docker Hub, 22.0.0 doesn't exist on Docker Hub)

I'll keep digging.

**Comment by ps-jay** [comment]
I've used `tcpdump` to capture traffic on the Artifactory server, and in the case of v22.x.x versions of Renovate, there are no HTTP request sent to Artifactory.  Hence the code is erroring out before even talking to Artifactory.

Here's the config.js:

```
$ cat config.js
module.exports = {

        {
            'datasources': ['docker'],
            'packagePatterns': ['renovate/*', 'rhel/*', 'centos/*', 'base/*'],
            'versioning': 'loose',
            'allowedVersions': '!/-(alpha|branch|dev)\..*/',
        },
        //--------------------------------------------------

        // Make use of Artifactory as a proxy to the real world
        {
            'managers': [
                'meteor',
                'npm',
            ],
            'registryUrls': [
                'https://artifactory.example.org/artifactory/api/npm/npm/',
            ]
        },
        {
            'managers': [
                'docker-compose',
                'dockerfile',
            ],
            'registryUrls': [
                'https://artifactory.example.org:6555',
            ]
        },
        {
            'managers': [
                'gomod',
            ],
            'registryUrls': [
                'https://artifactory.example.org/artifactory/api/go/go',
            ]
        },
    ],

    // npm is special
    npmrc: 'https://artifactory.example.org/artifactory/api/npm/npm/',

    // Renovate configuration below ------------------------

    platform: 'bitbucket-server',
    endpoint: 'https://stash.example.org/',
    autodiscover: 'true',

    rangeStrategy: 'bump',
    binarySource: 'docker',
    lockFileMaintenance: {
        enabled: true,
    },

};
```

**Comment by ps-jay** [comment]
Okay, now I've pasted our `config.js`, I got suspicious about this snippet:

```
        {
            'managers': [
                'docker-compose',
                'dockerfile',
            ],
            'registryUrls': [
                'https://artifactory.example.org:6555',
            ]
        },
```

Removing that block, fixes the fault I was observing.

The intention of that snippet of config was to try and proxy _all images_ via Artifactory - but I don't think that's working - nor is it really essential anyway.  So I shall remove it!

Thanks for looking @zharinov and @rarkins.  Sorry for the misdirection.
